ssz/src/main/java/org/ethereum/beacon/util/ssz/SSZSerializer.java

```diff
@@ -0,0 +1,333 @@
+        throw new RuntimeException(error);
+      } else {
+        result = setterAttempt.getValue();
+      }
+    } else {
+      result = constructorAttempt.getValue();
+    }
+
+    return result;
+  }
+
+  private SSZEncoderDecoder resolveEncoderDecoder(SSZScheme.SSZField field) {
+    SSZEncoderDecoder decoder;
+    if (registeredClassHandlers.containsKey(field.type)) {
+      decoder = registeredClassHandlers.get(field.type);
+    } else {
+      decoder = registeredClassHandlers.get(Object.class); // Container
+    }
+
+    return decoder;
+  }
+
+  private static Pair<Boolean, Object> createInstanceWithConstructor(Class clazz, Class[] params,
+                                                                     Object[] values) {
+    // Find constructor for params
+    Constructor constructor;
+    try {
+      constructor = clazz.getConstructor(params);
+    } catch (NoSuchMethodException e) {
+      return new Pair<>(false, null);
+    }
+
+    // Invoke constructor using values as params
+    Object result;
+    try {
+      result = constructor.newInstance(values);
+    } catch (Exception e) {
+      return new Pair<>(false, null);
+    }
+
+    return new Pair<>(true, result);
+  }
+
+  private static Pair<Boolean, Object> createInstanceWithSetters(
+      Class clazz, List<SSZScheme.SSZField> fields, Object[] values) {
+    // Find constructor with no params
+    Constructor constructor;
+    try {
+      constructor = clazz.getConstructor();
+    } catch (NoSuchMethodException e) {
+      return new Pair<>(false, null);
+    }
+
+    // Create empty instance
+    Object result;
+    try {
+      result = constructor.newInstance();
+    } catch (Exception e) {
+      return new Pair<>(false, null);
+    }
+
+    Map<String, Method> fieldSetters = new HashMap<>();
+    try {
+      for (PropertyDescriptor pd: Introspector.getBeanInfo(clazz).getPropertyDescriptors()) {
+        fieldSetters.put(pd.getName(), pd.getWriteMethod());
+      }
+    } catch (IntrospectionException e) {
+      String error = String.format("Couldn't enumerate all setters in class %s", clazz.getName());
+      throw new RuntimeException(error, e);
+    }
+
+    // Fill up field by field
+    for (int i = 0; i < fields.size(); ++i) {
+      SSZScheme.SSZField currentField = fields.get(i);
+      try {   // Try to set by field assignment
+        clazz.getField(currentField.name).set(result, values[i]);
+      } catch (Exception e) {
+        try {    // Try to set using setter
+          fieldSetters.get(currentField.name).invoke(result, values[i]);
+        } catch (Exception ex) {    // Cannot set the field
+          return new Pair<>(false, null);
+        }
+      }
+    }
+
+    return new Pair<>(true, result);
+  }
+
+  public void registerClassTypes(Set<Class> classTypes, SSZEncoderDecoder typeHandler) {
+    for (Class clazz : classTypes) {
+      if (registeredClassHandlers.put(clazz, typeHandler) != null) {
```

Comment:
This check makes it impossible to register a custom type which is represented with e.g. `byte[]` Java type (`byte[]` is already occupied by `BytesPrimitive`). Matching by Java type only could be insufficient in some cases. 

I think it would be better to match codecs by both Java and SSZ types and use latest registered codec instead of throwing an exception. This approach would allow client to customize primitives either.

Reply:
added